DatadogInternal/Sources/Swizzling/MethodSwizzler.swift

```diff
@@ -6,127 +6,160 @@
-        /// NOTE: RUMM-452 as we never add/remove methods/classes at runtime,
-        /// search operation doesn't have to wrapped in sync {...} although it's visible in the interface
-        var headKlass: AnyClass? = klass
-        while let someKlass = headKlass {
-            if let foundMethod = findMethod(with: selector, in: someKlass) {
-                return FoundMethod(method: foundMethod, klass: someKlass)
-            }
-            headKlass = class_getSuperclass(headKlass)
-        }
-        throw InternalError(description: "\(NSStringFromSelector(selector)) is not found in \(NSStringFromClass(klass))")
+/// Reference to type-erased override closure.
+private final class OverrideBox {
+    let closure: Any
+    init(_ closure: Any) {
+        self.closure = closure
     }
+}
 
-    public init() { }
+open class MethodSwizzler<Signature, Override> {
+    /// List of swizzling managed by this instance.
+    private var overrides: [(method: Method, `override`: OverrideBox)] = []
 
-    func originalImplementation(of found: FoundMethod) -> TypedIMP {
-        return sync {
-            let originalImp: IMP = implementationCache[found] ?? method_getImplementation(found.method)
-            return unsafeBitCast(originalImp, to: TypedIMP.self)
-        }
-    }
+    public init() { }
 
-    public func swizzle(
-        _ foundMethod: FoundMethod,
-        impProvider: (TypedIMP) -> TypedBlockIMP
-    ) {
-        sync {
-            let currentIMP = method_getImplementation(foundMethod.method)
-            let current_typedIMP = unsafeBitCast(currentIMP, to: TypedIMP.self)
-            let newImpBlock: TypedBlockIMP = impProvider(current_typedIMP)
-            let newImp: IMP = imp_implementationWithBlock(newImpBlock)
-
-            set(newIMP: newImp, for: foundMethod)
-
-            #if DD_SDK_COMPILED_FOR_TESTING
-            Swizzling.activeSwizzlingNames.append(foundMethod.swizzlingName)
-            #endif
+    /// Swizzle a method with a closure.
+    ///
+    /// - Parameters:
+    ///   - method: The method pointer to swizzle.
+    ///   - override: The closure to apply.
+    public func swizzle(_ method: Method, override: @escaping (Signature) -> Override) {
+        Swizzling.sync { swizzlings in
+            let org_imp = method_getImplementation(method)
+            let org = unsafeBitCast(org_imp, to: Signature.self)
+            let ovr: Override = override(org)
+            let ovr_imp: IMP = imp_implementationWithBlock(ovr)
+
+            let override = OverrideBox(override)
+            overrides.append((method, override))
+
+            swizzlings[method] = MethodSwizzling(
+                origin: org_imp,
+                override: override,
+                parent: swizzlings[method]
+            )
+
+            method_setImplementation(method, ovr_imp)
         }
     }
 
     /// Removes swizzling and resets the method to its original implementation.
     public func unswizzle() {
```

Comment:
I had hard time understanding this, IMO it would be cleaner to read if implemented without recursion.

When you unswizzle, you pop the stack until you find the target, and then call the swizzle method again on the type.

Reply:
It was even harder to implement 😄 I agree on readability, I will see if I can avoid recursion!

Reply:
Sounds good, happy with current set of tests.